sw/device/tests/crypto/cryptotest/firmware/ibex_sca.c

```diff
@@ -7,35 +7,186 @@
+
+// Generate Fixed vs Random (FvsR) values used by the test by calling the SCA
+// A random value is used to determine whether the fixed value is taken or a
+// random one is generated.
+static void generate_fvsr(uint32_t num_iterations, uint32_t fixed_data,
+                          uint32_t values[]) {
+  bool sample_fixed = true;
+  for (int i = 0; i < num_iterations; i++) {
+    if (sample_fixed) {
+      values[i] = fixed_data;
+    } else {
+      values[i] = prng_rand_uint32();
+    }
+    sample_fixed = prng_rand_uint32() & 0x1;
+  }
+}
+
+// Generate random values used by the test by calling the SCA PRNG.
+static void generate_random(uint32_t num_iterations, uint32_t values[]) {
+  for (int i = 0; i < num_iterations; i++) {
+    values[i] = prng_rand_uint32();
+  }
+}
+
+status_t handle_ibex_sca_tl_write_batch_fvsr_fix_address(ujson_t *uj) {
+  // Get number of iterations and fixed data.
+  ibex_sca_test_fvsr_t uj_data;
+  TRY(ujson_deserialize_ibex_sca_test_fvsr_t(uj, &uj_data));
+  TRY_CHECK(uj_data.num_iterations < 256);
+
+  // Generate FvsR values.
+  uint32_t values[256];
+  generate_fvsr(uj_data.num_iterations, uj_data.fixed_data, values);
+
+  // Get address of buffer located in SRAM.
+  uintptr_t sram_main_buffer_addr = (uintptr_t)&sram_main_buffer_batch;
+  mmio_region_t sram_region_main_addr =
+      mmio_region_from_addr(sram_main_buffer_addr);
+
+  // SCA code target.
+  for (int it = 0; it < uj_data.num_iterations; it++) {
+    sca_set_trigger_high();
+    // Give the trigger time to rise.
+    asm volatile(NOP100);
+    // Write random data into SRAM at the first address.
+    mmio_region_write32(sram_region_main_addr, 0, values[it]);
+    sca_set_trigger_low();
+    asm volatile(NOP100);
+  }
+
+  // Write back last value written into SRAM to validate generated data.
+  ibex_sca_result_t uj_output;
+  uj_output.result = values[uj_data.num_iterations - 1];
+  RESP_OK(ujson_serialize_ibex_sca_result_t, uj, &uj_output);
+  return OK_STATUS();
+}
+
+status_t handle_ibex_sca_tl_write_batch_fvsr(ujson_t *uj) {
+  // Get number of iterations and fixed data.
+  ibex_sca_test_fvsr_t uj_data;
+  TRY(ujson_deserialize_ibex_sca_test_fvsr_t(uj, &uj_data));
+  TRY_CHECK(uj_data.num_iterations < 256);
+
+  // Generate FvsR values.
+  uint32_t values[256];
+  generate_fvsr(uj_data.num_iterations, uj_data.fixed_data, values);
+
+  // Get address of buffer located in SRAM.
+  uintptr_t sram_main_buffer_addr = (uintptr_t)&sram_main_buffer_batch;
+  mmio_region_t sram_region_main_addr =
+      mmio_region_from_addr(sram_main_buffer_addr);
+
+  // SCA code target.
+  for (int it = 0; it < uj_data.num_iterations; it++) {
+    sca_set_trigger_high();
+    // Give the trigger time to rise.
+    asm volatile(NOP100);
+    // Write random data into SRAM.
+    mmio_region_write32(sram_region_main_addr, it * (ptrdiff_t)sizeof(uint32_t),
+                        values[it]);
+    sca_set_trigger_low();
+    asm volatile(NOP100);
+  }
```

Comment:
this code snippet, where the pattern is:
1. loop
2. set trigger high
3. NOP sled
4. memory write
5. set trigger low
6. NOP sled

could be made into a function I think? (again DRY principle here). The only thing that changes is the memory write?

Reply:
I see your point regarding code repetition but from a SCA point of view I would prefer not moving this in a function. The reason for that is that (i) the I might need to tweak the NOP sled for certain measurements (ii) want to not have function calls or conditional branches differentiating between the current test to have a static execution time (iii) and to keep execution time between the trigger as short as possible to save data captured on the oscilloscope.

Reply:
I see, would a configurable macro help then? Just some thoughts to help you increase the maintainability of this code. Though if you are planning on hand tweaking each function then perhaps it is fine as is.

Reply:
As we probably will need to manually tweak the function during the SCA testing, I would prefer leaving it as it is now.